src/lyra/abstract_domains/string/character_domain.py

```diff
@@ -0,0 +1,263 @@
+        set and must contain all characters in **certainly**. Therefore,
+        the **certainly** set is a subset of the **maybe** set.
+
+    """
+
+    class ConditionEvaluator(ExpressionVisitor):
+
+        def visit_Literal(self, expr, state=None, eval=None):
+            if expr.typ == StringLyraType():
+                return CharacterLattice(set(expr.val), set(expr.val))
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Input(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_VariableIdentifier(self, expr, state=None, eval=None):
+            return state.store[expr]
+
+        def visit_LengthIdentifier(self, expr):
+            pass
+
+        def visit_ListDisplay(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Range(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_AttributeReference(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Subscription(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Slicing(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_UnaryArithmeticOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_UnaryBooleanOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_BinaryArithmeticOperation(self, expr, state=None, eval=None):
+            if expr.operator == BinaryArithmeticOperation.Operator.Add:
+                left = self.visit(expr.left, state, eval)
+                right = self.visit(expr.right, state, eval)
+                return left.concat(right)
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_BinaryBooleanOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_BinaryComparisonOperation(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+    class Refinement(ExpressionVisitor):
+
+        def visit_Literal(self, expr: 'Literal', state=None, evaluation=None):
+            pass
```

Comment:
I expect the refinement to return state by default if there is nothing to do, not to pass (see the implementation of the interval domain).